gitlab/resource_gitlab_project.go

```diff
@@ -686,28 +686,42 @@ func resourceGitlabProjectDelete(d *schema.ResourceData, meta interface{}) error
 
 	// Wait for the project to be deleted.
 	// Deleting a project in gitlab is async.
+	// There is also a failure mode where, after a delete, a GET project can return a 404, and then
+	// repeatedly a 200 afterward. So we must retry the DELETE and wait for a consistent response.
 	stateConf := &resource.StateChangeConf{
 		Pending: []string{"Deleting"},
 		Target:  []string{"Deleted"},
 		Refresh: func() (interface{}, string, error) {
+			// Check that the project was deleted or marked for deletion.
 			out, response, err := client.Projects.GetProject(d.Id(), nil)
+			if response != nil && response.StatusCode == http.StatusNotFound {
+				return "", "Deleted", nil
+			}
 			if err != nil {
-				if response.StatusCode == 404 {
-					return out, "Deleted", nil
-				}
-				log.Printf("[ERROR] Received error: %#v", err)
-				return out, "Error", err
+				return "", "Error", err
 			}
 			if out.MarkedForDeletionAt != nil {
 				// Represents a Gitlab EE soft-delete
-				return out, "Deleted", nil
+				return "", "Deleted", nil
+			}
+
+			// If it is not deleted, try deleting it again.
+			response, err = client.Projects.DeleteProject(d.Id())
```

Comment:
Why??

Reply:
Heh. I left details in the #569 issue comments. Sometimes a deleted project becomes un-deleted. That's what I observed, though I am not sure the root cause.